### ISSUE TITLE
Switch with associated data instead of constructor methods for partials

### ISSUE THREAD

**Issue by ermik** [apollographql/apollo-ios#1260]
This is a feature request.

Given a `Node` interface type and `A`, `B`, and `C` types that conform to it, it the following schema:
```graphql
interface Node {
    id: ID!
}

type A implements Node {
}

```

**Comment by designatednerd** [comment]
It's a little hard to do it with the associated types since each type is different per query to make sure you're only trying to use the specific type requested in that query. However, I am working on something similar in [this PR for the Swift Codegen project](https://github.com/apollographql/apollo-ios/pull/1242) that will at least let you switch on the `__typename`.

**Comment by ermik** [comment]
You're right. I have overlooked the query-tie-in! in context of a query, the interface types remain a `switch`-able condition. It's just that `Node` type wouldn't be global:

```swift
struct NodeQuery {
    enum Node {
        case A(data: NodeQuery.StructA)
    }
}
```

and its cases would follow the same definition logic currently used for the `asType` getters.

`__typename` addition sounds like fun and I can't wait to use it!

**Comment by ermik** [comment]
I do see how the `Node` type appearing in different branches of the query would produce a different struct type in the associated value signature. Took me a second to get there. You're right.

**Comment by designatednerd** [comment]
🤔 I'm gonna do some noodling on this and see what might be doable - I like where you're going with it though.

**Comment by solidcell** [comment]
@designatednerd any updates on this? Not being able to exhaustively switch over union types adds a fair amount of fragility that's hard to ignore.

**Comment by AnthonyMDev** [comment]
We are in progress on a full re-write of Codegen for a 1.0 release. I've got some notes around this in the proposal here. https://github.com/apollographql/apollo-ios/blob/release/1.0/CodegenProposal.md#concrete-subtypes-as-enums

**Comment by mccarron** [comment]
@calvincestari Bump to this old issue, but is this still on a roadmap for 1.X? Would be super useful for something we're building.

**Comment by calvincestari** [comment]
Hi @mccarron - it's still not on our current roadmap unfortunately. @AnthonyMDev and I will discuss it but at this point it's unlikely to be a 1.x feature.
